### ISSUE TITLE
cannot load LLaVA model onto GPU on Colab

### ISSUE THREAD

**Issue by DiTo97** [abetlen/llama-cpp-python#1027]
Hi @abetlen,

we have been trying to run LLaVA 1.5 using the LLaMA.cpp engine on Colab with a T4 GPU.

The package has been installed using the following parameters:
```shell
CMAKE_ARGS="-DLLAMA_BLAS=ON -DLLAMA_BLAS_VENDOR=OpenBLAS" python -m pip install llama-cpp-python
```

The model runs correctly, but it always sticks to the CPU even when setting `n_gpu_layers=-1` as seen in the docs.

Hereafter, I will paste relevant snippets of code with the memory and compute usage:
```python
from llama_cpp import Llama
from llama_cpp.llama_chat_format import Llava15ChatHandler

handler = Llava15ChatHandler(clip_model_path="mmproj-model-f16.gguf", verbose=True)

model = Llama(
    model_path="ggml-model-q4_k.gguf",
    chat_handler=handler,
    n_ctx=2048,
    logits_all=True,
    n_gpu_layers=-1
)

model.create_chat_completion(
    messages = [
        {"role": "system", "content": "..."},
        {
            "role": "user",
            "content": [
                {"type": "image_url", "image_url": {"url": "..."}},
                {"type" : "text", "text": "..."}
            ]
        }
    ]
)
```

![memory-and-compute-usage](https://github.com/abetlen/llama-cpp-python/assets/40361744/174c8ab7-53e0-432b-8327-447ddf7ecc83)

The model weights have been downloaded from the [Hugging Face repository](https://huggingface.co/mys/ggml_llava-v1.5-7b/tree/main).

**Comment by abetlen** [comment]
@DiTo97 your install script should install with CUBLAS support not openblas to take advantage of the GPU. Note however that the clip model itself will always run on CPU as the current implementatoin in llama.cpp does not allow for offloading.

**Comment by DiTo97** [comment]
Thanks @abetlen.

Are there any future plans for enabling CLIP offloading that you are aware of?

@Fra-Tsuna, see if it works with CUBLAS.

**Comment by abetlen** [comment]
@DiTo97 afaik no work yet in llama.cpp to support this.

I do have an alternative solution that might save some time by leveraging 3rd party python libraries for this but don't have time at the moment to implement it

- Create a new LlavaChatHandler implementation in `llama_chat_format.py` with an almost identical implementation to the CPU version
- Instead of loading a llama.cpp clip model and calling `llava_image_embed_make_with_bytes` we instead load the model with a Pytorch or JAX implementation that makes use of the GPU
- Create a new `llava_image_embed` c structure that holds the output of the custom implementation and pass it to `llava_eval_image_embed` like normal

**Comment by hockeybro12** [comment]
@DiTo97  did you end up offloading CLIP to GPU?

**Comment by DiTo97** [comment]
@hockeybro12, nope. In the end we looked into other VLMs as their performance is a bit better on our experiments than LLaVA.

still using LLaMA.cpp for LLMs, but no CLIP is involved

**Comment by hockeybro12** [comment]
@DiTo97 what are you using instead?

**Comment by abetlen** [comment]
@DiTo97 @hockeybro12 just a heads up CLIP now supports gpu offloading at least for CUDA
